pkg/storage/stores/tsdb/head_manager.go

```diff
@@ -131,10 +140,47 @@ func NewHeadManager(logger log.Logger, dir string, metrics *Metrics, tsdbManager
 
 	})
 
+	go m.loop()
+
 	return m
 }
 
+func (m *HeadManager) loop() {
+	m.loopWg.Add(1)
+	defer m.loopWg.Done()
+
+	ticker := time.NewTicker(defaultBuildRotatedHeadsInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			m.mtx.Lock()
+			if m.prev != nil {
+				if err := m.buildTSDBFromWAL(m.prev.initialized); err != nil {
+					level.Error(m.log).Log(
+						"msg", "failed building tsdb from rotated out period",
+						"period", m.period.PeriodFor(m.prev.initialized),
+						"err", err,
+					)
+					// keeps retrying on tick until it gets rotated out
+					// should we be worried about this?
```

Comment:
This would at least result in a different behaviour than there is now. I guess if building the TSDB once, it would very likely fail with the same error again in subsequent iterations.

Reply:
agree, currently this would result in `Append` failure
Not sure how likely this is to fail, should I add a metric instead for some visibility?

Reply:
curious isn't `m.prev != nil` takes care of `Append()` failure? before building TSDB from WAL?

Reply:
not sure if I understood the question correctly :/

actually the next `Append` won't even call the `Rotate` method since the new head covers that period.
So without this PR. we would only try to build prev head once (as part of `Append`), if that fails we leave it for recovery from WAL during start-up